contrib/autopilot/lib_autopilot.py

```diff
@@ -0,0 +1,430 @@
+
+    def calculate_proposed_channel_capacities(self, pdf, balance=1000000):
+        minimal_channel_balance = 20000  # lnd uses 20k satoshi which seems reasonble
+
+        min_probability = min(pdf.values())
+        needed_total_balance = math.ceil(
+            minimal_channel_balance / min_probability)
+        self.__logger.info(
+            "Need at least a balance of {} satoshi to open {} channels".format(
+                needed_total_balance, len(pdf)))
+        while needed_total_balance > balance and len(pdf) > 1:
+            min_val = min(pdf.values())
+            k = [k for k, v in pdf.items() if v == min_val][0]
+            self.__logger.info(
+                "Not enough balance to open {} channels. Remove node: {} and rebalance pdf for channel balances".format(
+                    len(pdf), k))
+            del pdf[k]
+
+            s = sum(pdf.values())
+            pdf = {k: v / s for k, v in pdf.items()}
+
+            min_probability = min(pdf.values())
+            needed_total_balance = math.ceil(
+                minimal_channel_balance / min_probability)
+            self.__logger.info(
+                "Need at least a balance of {} satoshi to open {} channels".format(
+                    needed_total_balance, len(pdf)))
+
+        return pdf
+
+
+
+    def find_candidates(self, num_items=21,strategy = Strategy.DIVERSE, 
+                        percentile = None):
+        self.__logger.info("running the autopilot on a graph with {} nodes and {} edges.".format(
+            len(self.G.nodes()), len(self.G.edges())))
+        """
+        Generates candidates with several strategies
+        """
+        sub_k = math.ceil(num_items / 4)
+        self.__logger.info(
+            "GENERATE CANDIDATES: Try to generate up to {} nodes with 4 strategies: (random, central, network Improvement, liquidity)".format(num_items))
+        # FIXME: should remember from where nodes are known
+        
+        res = self.__create_pdfs()
+        
+        candidats = set()
+        # FIXME: Run simulations to decide the following problem:
+        """
+        we can either do a global sampling by merging all probability 
+        distributions and sample once from them or we can sample from 
+        each probability distribution and merge the results. These processes
+        are obviously not commutative and we need to check which one seems
+        more reasonable.
+        My (renepickhardt) guts feeling says several samples which are 
+        merged gives the best of all worlds where the other method would 
+        probably result in something that is either pretty uniform or 
+        dominated by one very skew distribution. as mentioned this needs
+        to be tested
+        """
+        if strategy == Strategy.DIVERSE:
+            for strategy, pdf in res.items():
+                tmp = self.__sample_from_percentile(pdf, percentile, sub_k)
+                candidats = candidats.union(set(tmp))
```

Comment:
This is very likely to result in duplicates: central nodes may also be rich. This then result in less than the expected number of channels being opened.

Reply:
yes and no. the likelihood of duplicates drops tremendously when sampling from the full probability distribution. 
My current Idea was to say if duplicates emerge we have less channels. this could of course easily be fixed by just checking the size of the result set and polling channels again. In the best case polling with one of the strategies from which the duplicates emerged.